privacyidea/api/before_after.py

```diff
@@ -141,6 +141,12 @@ def before_request():
     # from the Form data or from JSON in the request body.
     update_config_object()
     request.all_data = get_all_params(request.values, request.data)
+    if g.logged_in_user.get("role") == "user":
+        # A user is calling this API. First thing we do is restricting the user parameter.
+        # ...to restrict token view, audit view or token actions.
+        request.all_data["user"] = g.logged_in_user.get("username")
+        request.all_data["realm"] = g.logged_in_user.get("realm")
```

Comment:
Prior to this change, ``request.all_data["resolver"]`` was set to the logged-in user's resolver if the current user's role was ``user``. This is not the case anymore, I imagine this could cause problems in other code.

Reply:
I moved the whole user thingy up. This is *only* for the case, if the user is logged in.
The thing is, that the ``g.logged_in_user`` does not contain a resolver!

Right after this code you see:

~~~python
    if g.logged_in_user.get("role") == "user":
        # A user is calling this API. First thing we do is restricting the user parameter.
        # ...to restrict token view, audit view or token actions.
        request.all_data["user"] = g.logged_in_user.get("username")
        request.all_data["realm"] = g.logged_in_user.get("realm")

    try:
        request.User = get_user_from_param(request.all_data)
~~~
This is the important part. ``request.User`` now contains a complete User Object including the resolver!
And after this we only work with ``request.User``. So to my understanding this is totally fine.

Reply:
I think this does not need to be change. Otherwise take a look at  916c01da

Reply:
This must not be a problem!
Policies should/must not use the resolver information from the request data! Because usually the resolver is not contained in the request data.

E.g. an admin enrolls a token to a *user* "friedrich" in *realm* "netknights". Not a *user* "friedrich" in *resolver* "developers" in *realm* "netknights". If this worked before, this was a bad design! ;-)

This is why initially we set a user object in ``request.User``. And in an ideal world imho this should be used afterwars. (Even if it would not always be the case). 
So I would like to leave it that way, even if we change the behaviour at certain places. But this would result in a bad design at the other places.

(And also, the ``logged_in_user`` also has no componente "resolver".)

Reply:
Yes. E.g. ``check_base_action`` actually uses the parameter "resolver".

Reply:
I think about it.

Reply:
I would suggest the following:

~~~~python
    request.all_data = get_all_params(request.values, request.data)
    if g.logged_in_user.get("role") == "user":
        # A user is calling this API. First thing we do is restricting the user parameter.
        # ...to restrict token view, audit view or token actions.
        request.all_data["user"] = g.logged_in_user.get("username")
        request.all_data["realm"] = g.logged_in_user.get("realm")

    try:
        request.User = get_user_from_param(request.all_data)
        # overwrite or set the resolver parameter in case of a logged in user
        if g.logged_in_user.get("role") == "user":  ### <<<<<< new
            request.all_data["resolver"] = request.User.resolver
    except AttributeError:
        # Some endpoints do not need users OR e.g. the setPolicy endpoint
        # takes a list as the userobject
        request.User = None
    except UserError:
        # In cases like the policy API, the parameter "user" is part of the
        # policy and will not resolve to a user object
        request.User = User()
~~~~
Only if the the role user is logged in, then we set the resolver. 
This might also acutally make sense, since it overwrites the resolver, in case a user would willingly set a wrong resolver. 
(Previously the resolver also was only set for users)

The admin is free to choose the resolver.

Reply:
:+1:

Reply:
:+1: